fw/http_parser.c

```diff
@@ -310,6 +310,32 @@ do {									\
 	__FSM_I_MOVE_BY_REF_n(to, 1, 0)
 
 #define __FSM_I_MOVE(to)		__FSM_I_MOVE_n(to, 1)
+
+#define __FSM_I_MOVE_nofixup_n(to, n)					\
+do {									\
+	p += n;								\
+	if (unlikely(__data_off(p) >= len)) {				\
+		parser->_i_st = &&to; /* start from state @to next time */\
+		__FSM_EXIT(TFW_POSTPONE);				\
+	}								\
+	goto to;							\
+} while (0)
+
+#define __FSM_I_MOVE_nofixup(to)	__FSM_I_MOVE_nofixup_n(to, 1)
+
+#define __FSM_I_MOVE_fixup_pos(to, n, pos, fixup_len, flag)		\
+do {									\
+	BUG_ON(n < 0);							\
+	parser->_i_st = &&to;						\
+	p += n;  							\
+	if (unlikely(__data_off(p) >= len)) {				\
+		__msg_hdr_chunk_fixup(pos, fixup_len);			\
+		__msg_chunk_flags(flag);				\
+		__FSM_EXIT(TFW_POSTPONE);				\
+	}								\
+	goto to;							\
+} while (0)
```

Comment:
I don't think we need to fixup each character in `Req_I_Forwarded_Host_Unquoted` and similar states (the test case like https://github.com/tempesta-tech/tempesta/pull/1629/files#diff-9b013f9b927ee84259765a331c57e94f2240a5eb38d7c9a10afada9aaec70050R3889 must be satisfied). Once we face `c` such as `!(isalnum(c) || c == '.' || c == '-')`, we fixup the host value - that's correct, but we should fixup the current chunk for `(isalnum(c) || c == '.' || c == '-')` only if we reach the end of data. We can open the current chunk  say in `Req_I_Forwarded_Host_Start`.

IIRC we can 'open' a chunk either with explicit `__msg_field_open()` or just using explicit fixup macros like `__FSM_I_MOVE_fixup` for all pieces of the header.

There is a comment https://github.com/tempesta-tech/tempesta/blob/master/fw/http_parser.c#L332 about this sophisticated topic - maybe it should be made more clear...

Reply:
Maybe I missed up something. Let's see:

`Req_I_Forwarded_Host_Unquoted` and similar consist of two parts.

In the first part 
```
if (likely(isalnum(c) || c == '.' || c == '-')) 
        __FSM_I_MOVE_flag(Req_I_Forwarded_Host_Unquoted, TFW_STR_VALUE);
```
we have "loop" which increments `@p` and jumps to `Req_I_Forwarded_Host_Unquoted` state again without fixup while data is available and @c is allowed symbol.
When incorrect symbol is reached we fall to second part where we fixup correct string by using `data + __fsm_sz` as beginning of data required to fixup:
Forwarded:
```
if (likely(p - (data + __fsm_sz))) {
            __msg_hdr_chunk_fixup(data + __fsm_sz, (p - (data + __fsm_sz)));
            __msg_chunk_flags(TFW_STR_VALUE);
        }
```

Regular Host:
```
if (p - data) {
            __msg_hdr_chunk_fixup(data, (p - data));
            __msg_chunk_flags(TFW_STR_VALUE);
        }
```

It's ok, and same method use now for parsing regular Host header. Except `data + __fsm_sz` part, there it's not needed with regular Host, because value always starts at `@data` beginning. But with Forwarded we must know at which position in `@data` we starts to parse value, we have multiple values(for= or host=). And as I mentioned before, `@p` increments but not fixuped, when we need to fixup(e.g whole string was checked) we also need to know where we start incrementing `@p`. We can't use `__msg_hdr_chunk_fixup(data, (p - data));` like in Host parsing, it will been fixuped from `@data` beginning, also we can't use `__FSM_I_MOVE_fixup` because it's starts fixup at the `@p`. Same problem with splitted requests, when data exhausted each \*MOVE\*(except \*MATCH_MOVE\* it uses function for validate value, which returns number of correct symbols without modifying `@p` and if needed postpone fixups `@p + returned size`) macro will fixup `@data + @len`, but it's wrong, some peace of data might have already been fixuped before in other state e.g "host=" part. 

`__FSM_I_MOVE_fixup_pos` introduced to avoid this, it allows to specify start position in `@data` pointer and length of necessary data on postpone, but in other cases it just MOVE with data bounds check.
`__FSM_I_MOVE_fixup_pos` it's same as `__FSM_I_MOVE_BY_REF_n` with two additional params, when request postponed `__FSM_I_MOVE_fixup_pos` starts to fixup not from `@data`, it will use passed value.

That's why I use `__fsm_sz = __data_off(p);` in `Req_I_Forwarded_Host_Start_Unquoted`. It's necessary to save length of already parsed data which is also start point of new parameter value.

Reply:
Yes, you're right: in other places we either don't have many chunks in the parsed header value or use some data eaters like the AVX alphabet matchers or `__parse_ulong()`.

However, one thing about fixups is that they must fixup all data. E.g. if we have `Some-header: some-garbage; interesing-value`, then we can not fixup only `interesing-value`, we must fixup the whole header value to be able to modify the underlying `skb` data. So the whole fixup logic was developed to fixup contigous chunks of data. You can fixup either the whole data from `data` pointer or from _previously_ fixed up chunk. Fixing up a chunk at some offset and of some length looks dangerous and introducing additional complexity (the fixup logic is already overcomplecated).

I think the code
```C
        __FSM_STATE(Req_I_Forwarded_Host_v6_Quoted) {
                /* See Req_UriAuthorityIPv6 processing. */
                if (likely(isxdigit(c) || c == ':'))
                        __FSM_I_MOVE_fixup_pos(Req_I_Forwarded_Host_v6_Quoted, 1,                                 
                                               data + __fsm_sz, p - (data + __fsm_sz),                            
                                               TFW_STR_VALUE);
                if (likely(c == ']')) {
                        __msg_hdr_chunk_fixup(data + __fsm_sz, p - (data + __fsm_sz) + 1);                        
                        __msg_chunk_flags(TFW_STR_VALUE);
                        __FSM_I_MOVE_fixup_pos(Req_I_Forwarded_Host_End_Quoted,                                   
                                               1, p - 1, 1, TFW_STR_VALUE);                                       
                }

                return CSTR_NEQ;
}
```
_can be_ rewrited as something like
```C
        __FSM_STATE(Req_I_Forwarded_Host_v6_Quoted) {
                /* See Req_UriAuthorityIPv6 processing. */
                __fsm_sz = 0;
                while (likely(isxdigit(*p) || *p == ':')) {
                        __fsm_sz++;
                        if (unlikely(!_data_available(p, __fsm_sz)))
                                __FSM_I_MOVE_fixup(Req_I_Forwarded_Host_v6_Quoted,
                                                   __fsm_sz, TFW_STR_VALUE);
                }
                if (likely(c == ']'))
                        __FSM_I_MOVE_fixup(Req_I_Forwarded_Host_End_Quoted,                                       
                                           __fsm_sz, TFW_STR_VALUE);

                return CSTR_NEQ;
        }
```
This loop is basically what we have in `__parse_ulong()` and the usage of the loop is like what we have in `Req_I_H_Port` state. The loop is attractive since it gets rid of the state transition logic, which is an overhead for simple spinning in the state.

Since you have several such places the code can be generalized with _ad-hoc_ macro (not a generic fixup macro). Moreover, we can embed the body of `__FSM_I_MOVE_fixup_f` into the macro, e.g. to get rid of the second check of available data `__data_off(p) >= len` and just do `_FSM_EXIT(TFW_POSTPONE)` if there is no available data.